internal/internal_task_handlers.go

```diff
@@ -1671,18 +1707,18 @@ func (i *cadenceInvoker) Heartbeat(details []byte, skipBatching bool) error {
 			var detailsToReport *[]byte
 
 			i.Lock()
-			detailsToReport = i.lastDetailsToReport
+			detailsToReport = i.detailsToReport
 			i.hbBatchEndTimer.Stop()
 			i.hbBatchEndTimer = nil
-			i.Unlock()
 
 			if detailsToReport != nil {
 				// TODO: there is a potential race condition here as the lock is released here and
-				// locked again in the Hearbeat() method. This possible that a heartbeat call from
+				// locked again in the Heartbeat() method. This possible that a heartbeat call from
 				// user activity grabs the lock first and calls internalHeartBeat before this
 				// batching goroutine, which means some activity progress will be lost.
-				i.Heartbeat(*detailsToReport, false)
+				i.heartbeatAndScheduleNextRun(*detailsToReport)
```

Comment:
Do we still have the race condition given that `Unlock()` has been moved after this call?

Reply:
No I don't think so. I'll remove this comment.